cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"encoding/json"
+	"net"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/golang/glog"
+	"github.com/vishvananda/netlink"
+)
+
+// Update iptables rules based on google cloud load balancer VIPS
+//
+// This is needed because the GCP L3 load balancer doesn't actually do DNAT;
+// the destination IP address is still the VIP. Normally, there is an agent that
+// adds the vip to the local routing table, tricking the kernel in to thinking
+// it's a local IP and allowing processes doing an accept(0.0.0.0) to receive
+// the packets. Clever.
+//
+// We don't do that. Instead, we DNAT with conntrack. This is so we don't break
+// existing connections when the vip is removed. This is useful for draining
+// connections - take ourselves out of the vip, but service existing conns.
+
+const (
+	// FIXME: support multiple interfaces
+	gcpMetadataForwardedIPs = "instance/network-interfaces/0/forwarded-ips/?recursive=true"
+	// name of nat chain for iptables masquerade rules
+	gcpMasqChainName      = "gcp-vips"
+	gcpLocalMasqChainName = "gcp-vips-local"
+)
+
+type gcpHandler struct {
+	vips      []string
+	ip4tables *iptables.IPTables
+	// GCP doesn't support IPv6
+	// ip6tables *iptables.IPTables
+	stopCh chan struct{}
+}
+
+var _ handler = &gcpHandler{}
+
+func newGCPHandler() (handler, error) {
+	ipt4, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(5),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcpHandler{
+		ip4tables: ipt4,
+		stopCh:    make(chan struct{}),
+	}, nil
+}
+
+// onFailure: either stop the routes service, or write downfile
+func (gcp *gcpHandler) onFailure() error {
+	select {
+	case gcp.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+	return gcp.deleteRules()
+}
+
+// onSuccess: either start routes service, or remove down file
+func (gcp *gcpHandler) onSuccess() error {
+	if err := gcp.syncRulesOnce(); err != nil {
+		return err
+	}
+
+	// bz1930457 delete stale conntrack entries to the apiserver
+	// FIXME: assume IPv4 only because GCP doesn't support IPv6
+	for _, vip := range gcp.vips {
+		// sanity check to not flush unexpected IPs
+		// provided by the metadata service
+		ip := net.ParseIP(vip)
+		if ip == nil || !ip.IsGlobalUnicast() {
+			continue
+		}
+		filter := &netlink.ConntrackFilter{}
```

Comment:
any states we can add here, just in case we over-match?

Reply:
this happens once, only on success and if the nated ip is the vip (that means it wasn't redirected)
It looks safe to me